### PR TITLE
Implement class type script

### PR DESCRIPTION
**Pull request by duanyytop**
- [x] Impl class cell creation action
- [x] Impl class cell updating action
- [x] Impl class cell destroying action
- [x] Add successful tests to class cell type script
- [x] Add error tests to class cell type script
- [x] Clean code

### CODE REVIEW

File: contracts/class-type/src/entry.rs [hunk 1]
```diff
@@ -1,6 +1,131 @@
...
+            Some(data) => Ok(Issuer::from_data(&data[..])?),
+            None => Err(Error::IssuerDataInvalid),
+        }?;
+    let output_issuer =
+        match load_cell_data_by_type_args(Source::Output, &check_issuer_args(class_args)) {
+            Some(data) => Ok(Issuer::from_data(&data[..])?),
+            None => Err(Error::IssuerDataInvalid),
+        }?;
+
+    if output_issuer.class_count < input_issuer.class_count {
```
Comment (chuijiaolianying) [inline contracts/class-type/src/entry.rs:66]:
```suggestion
    if output_issuer.class_count <= input_issuer.class_count {
```

Comment (duanyytop) [reply contracts/class-type/src/entry.rs:66]: No `=`. The `class_count` of input can be equal to the `class_count` of output.

File: contracts/script-utils/src/class.rs [hunk 2]
```diff
@@ -1 +1,93 @@
...
+        if version != 0 {
+            return Err(Error::VersionInvalid);
+        }
+
+        let mut total_buf = [0u8; 4];
+        let mut issued_buf = [0u8; 4];
+        total_buf.copy_from_slice(&data[1..5]);
+        issued_buf.copy_from_slice(&data[5..9]);
+        let total = u32::from_be_bytes(total_buf);
+        let issued = u32::from_be_bytes(issued_buf);
```
Comment (chuijiaolianying) [inline contracts/script-utils/src/class.rs:47]: Can we write a helper function to read buf and convert type?

Comment (duanyytop) [reply contracts/script-utils/src/class.rs:47]: Good idea!

Comment (duanyytop) [reply contracts/script-utils/src/class.rs]: Updated. https://github.com/nervina-labs/ckb-nft-scripts/pull/2/commits/1f00af215deed929755387ccb489754f1b01398a

**Review by chuijiaolianying — APPROVED** [review]
LGTM

File: contracts/script-utils/src/issuer.rs [hunk 3]
```diff
@@ -1,10 +1,16 @@
-use core::result::Result;
-
 use crate::error::Error;
+use crate::helper::{parse_dyn_vec_len, u32_from_slice};
+use core::result::Result;
 
-const ISSUER_MIN_LEN: usize = 11;
+const ISSUER_DATA_MIN_LEN: usize = 11;
 pub const ISSUER_TYPE_ARGS_LEN: usize = 20;
```
[the remark this answers is not in the document]
Comment (chuijiaolianying) [reply contracts/script-utils/src/issuer.rs:6]: In the [RFC](https://talk.nervos.org/t/rfc-multi-purpose-nft-draft-spec/5434):

> Here we define the IssuerID = issuer_cell_typescript_hash[:20] for the future usage.

File: contracts/script-utils/src/class.rs [hunk 4]
```diff
@@ -1 +1,89 @@
-pub const CLASS_TYPE_ARGS_LEN: usize = 52;
+use crate::error::Error;
+use crate::helper::{parse_dyn_vec_len, u32_from_slice};
+use alloc::vec::Vec;
+use core::result::Result;
+
+const CLASS_DATA_MIN_LEN: usize = 16;
```
[the remark this answers is not in the document]
Comment (duanyytop) [reply contracts/script-utils/src/class.rs:9]: Updated. https://github.com/nervina-labs/ckb-nft-scripts/pull/2/commits/37d78afddb691277e52b162532891e6186ae66f3

File: contracts/script-utils/src/class.rs [hunk 5]
```diff
@@ -1 +1,89 @@
...
+        let issued = u32_from_slice(&data[5..9]);
+
+        if total > 0 && issued >= total {
+            return Err(Error::ClassTotalSmallerThanIssued);
+        }
+
+        let configure: u8 = data[9];
+
+        let name_len = parse_dyn_vec_len(&data[10..12]);
+        if data.len() < name_len + 12 {
```
[the remark this answers is not in the document]
Comment (duanyytop) [reply contracts/script-utils/src/class.rs]: Updated. https://github.com/nervina-labs/ckb-nft-scripts/pull/2/commits/37d78afddb691277e52b162532891e6186ae66f3

**Review by shaojunda — APPROVED** [review]

**Pull request merged**